extract subclassess_supporting

kbrock opened the pull request:
Provide a mechanism that allows a class (e.g.: `Vm`) to easily generate a `:types =>` clause that will only return instances that support a particular feature

This was very cool in ems, and looks like it could provide some nice features for other classes.

This is here to build upon https://github.com/ManageIQ/manageiq-ui-classic/pull/7954#discussion_r746862439

kbrock commented on app/models/ext_management_system.rb:
```diff
@@ -50,23 +46,19 @@ def self.supported_types_for_create
 
   def self.entities_for_label_mapping
-    label_mapping_classes.reduce({}) { |all_mappings, klass| all_mappings.merge(klass.entities_for_label_mapping) }
+    subclasses_supporting(:label_mapping).reduce({}) { |all_mappings, klass| all_mappings.merge(klass.entities_for_label_mapping) }
```
would this work  does it read better?
```suggestion
    subclasses_supporting(:label_mapping).flat_map(&:entities_for_label_mapping)
```

agrare replied:
Yeah `.flat_map` is probably how I would have written it

Fryguy replied:
Those return two different things...one creates a hash and the other creates an array.

kbrock replied:
dumb idea, who would suggest this?

(aka sorry)

kbrock commented on app/models/zone.rb:
```diff
@@ -197,8 +197,7 @@ def ems_clouds
   def ems_metrics_collectable
-    supported_ems_types = ExtManagementSystem.subclasses_supports?(:metrics).map(&:name)
-    ext_management_systems.where(:type => supported_ems_types).select { |e| e.supports?(:metrics) }
+    ext_management_systems.supporting(:metrics)
```
https://github.com/ManageIQ/manageiq/pull/21553

agrare replied:
Nice!

agrare replied:
We discussed this one on slack, but we still need to evaluate the instance-level supports blocks (e.g. azure class supports metrics but instances can not support it due to insights service not being enabled)

agrare replied:
I wonder if we should formalize the "class-level" supports and the "instance-level" supports.  I'm a little concerned that we're changing the default behavior here from instance-level to class-level and we might end up introducing bugs like this almost was.  cc @Fryguy

kbrock replied:
Well, it feel to me like we have a kind of validation embedded in this supports concept.
Does this ems type support this concept (class level) and does this ems in particular support this feature (instance level)
It gets confusing since we only sometimes implement instance level checking.

agrare replied:
Yeah we have to be very careful about changing something from:
```ruby
Class.all.select { |obj| obj.supports?(:feature) }
```
and
```ruby
Class.descendants.select { |klass| klass.supports?(:feature) }
```

One only checks the class supports it, the other checks that the instance supports it and they aren't interchangeable

agrare commented:
`subclasses_supporting(:feature)` does read nicely compared to `subclasses_supports?(feature)`

agrare commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -116,6 +116,30 @@ def supports?(feature)
     end
 
+    # all subclasses that are considered for supporting features
+    def supported_subclasses
```
The idea of "supported_subclasses" might be unique to ExtManagementSystem, we could go with `descendants.select {}` and just override that in ExtManagementSystem.

Though thinking more about it this would give other classes a place to override if they want/need to....

Fryguy replied:
I'm concerned this will create mental conflict with overloading the word "support".  In EMS, supported means "subclasses that are permitted to be displayed" which is different than "supports feature", which is about the features.  Adding the method here will only create more confusion.

I'm wondering if we should come up with a new term on the EMS side and called the existing `supported_subclasses` something else...maybe `permitted_subclasses`, since it's tied to the Vmdb::PermissionsStore

Fryguy replied:
Opened https://github.com/ManageIQ/manageiq/pull/21556, which renames supported_subclasses to permitted_subclasses.  Once that's in then, then it unlocks a method named `supported_subclasses` for other use, such as this one here.

Fryguy replied:
#21556 is merged...does that now change your comment here @agrare ?

agrare replied:
Yeah I'm good with this name now that it isn't conflicting with what is in ExtManagementSystem, we'll just need to define `supported_subclasses` as `permitted_subclasses` in `ExtManagementSystem` now

agrare commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -116,6 +116,30 @@ def supports?(feature)
+    # Ems types (class names) that support this feature
+    # Ems.where(:type => Host.ems_types_supporting(:create))
+    def ems_types_supporting(feature)
+      subclasses_supporting(feature).map(&:module_parent).map(&:name)
```
This is perfect for querying on `.where(:type =>)` but we might sometimes want a list of the class names, could you split this up into one that returns the class names then this becomes `emses_supporting(feature).map(&:name)`

(plural emses doesn't look great but :shrug: )

chessbyte replied:
should we name it `providers_supporting` ?

Fryguy replied:
@agrare This PR now has the `def supporting(feature)` method which returns a scope - is that what you were looking for ?

agrare replied:
Well that returns instances of that class where this one is about returning the EMS classes.

Fryguy replied:
Ohhh I see....yeah then I agree with you.

agrare replied:
And actually a scope that returns the EMS instances (similar to `.supporting(feature)`) would make these: https://github.com/ManageIQ/manageiq-ui-classic/pull/7958/files#diff-b81586b5ee3afef460d25dce2deda502148e37faba61591f771df88c23128daeR32 even simpler

agrare replied:
`ManageIQ::Providers::CloudManager::AuthKeyPair.providers_supporting(:create)` for example

kbrock replied:
I added the providers_supporting that returns classes, but to be honest, I do not see us needing types_supporting as I think `supporting()` will meet our needs.

I also do not think we need `providers_supporting`. Wonder if we want to convert `provider_types_supporting` into a scope and we could also do a scope instead. merge would add to another chain.

I guess having strings is good because we store stuff in the session

agrare replied:
Yeah I like the scope over something that returns the class name strings just to then pass that to a `.where(:type =>`

kbrock replied:
@agrare drop `type_supporting`?

Thinking `provider_type_supporting` may stay for a little until we get a good grasp of it.

kbrock replied:
ok, added `providers_supporting`
This removes our need for looking at provider types.
if we need part of this we can scope merge it

Fryguy commented:
Merged https://github.com/ManageIQ/manageiq-api/pull/1098 and dependents.  This can be updated to account for those changes, or the can be done in a follow up.

Fryguy reviewed (changes requested):
Not against this PR, but blocking merge as I'd like https://github.com/ManageIQ/manageiq/pull/21556 to go in first.

Fryguy commented:
#21556 is merged which will probably create conflicts here.

Fryguy commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -116,6 +116,30 @@ def supports?(feature)
+      supported_subclasses.select { |subclass| subclass.supports?(feature) }
+    end
+
+    # scope to query all those classes that support a particular feature
+    def supporting(feature)
+      where(:type => types_supporting(feature))
+    end
```
This is so nice. 😍

Fryguy commented on spec/models/mixins/supports_feature_mixin_spec.rb:
```diff
@@ -200,4 +205,56 @@ def initialize(options = {})
     end
   end
+
+  describe ".subclasses_supporting" do
```
Need to a spec for `.supporting`

kbrock replied:
getting our caching of descendants to not blow up between tests was the hardest part of this code.
But I'm pretty sure we got it working.

kbrock commented:
pushed

hah. This is a good one.

We define an ems in the specs. And when we leave, the ems goes away.
**But** `ActiveSupport::DescendantTracker` caches this ems as a child of `ExtManagementSystem`. So this ems is returned for future tests.

Symptom: `Ems.descendants.map(&:default_blacklist_names)` found classes that did not define that method. (because, yes, we did not define that method). Solution is to remove the entry from the cache rather than let this not freak out.

Fryguy commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -116,6 +116,33 @@ def supports?(feature)
+      permitted_subclasses.select { |subclass| subclass.supports?(feature) }
+    end
+
+    # Provider classes that support this feature
+    def provider_classes_supporting(feature)
+      subclasses_supporting(feature).map(&:module_parent)
+    end
```
I'm not sure what the "provider" part of this method name means now (unless it's assuming that supports feature will only be mixed into "provider" related things (e.g. availability_zone, instances, etc), but _not_ non-provider things)

kbrock replied:
this is "give me the ems classes that support this feature"

kbrock replied:
This is the class version of the similar `providers_supporting` (which I think is more useful)

example:

Give me the Ems where I can create a network router:

```
  NetworkRouter.provider_supporting(:create) # => [Ems]
```

Fryguy commented on app/models/mixins/supports_feature_mixin.rb:
```diff
@@ -116,6 +116,33 @@ def supports?(feature)
     end
 
+    # all subclasses that are considered for supporting features
+    def permitted_subclasses
```
We just have to be careful to not override the one defined in EMS

kbrock replied:
we need the concept of all subclasses that are considered for supporting features. In Ems this is called `permitted_subclasses`. This is the method used by the ems version of this code that has been extracted to here.

Are you thinking we want a separate concept, or just concerned that we would override this method for EMS (instead of ensuring that EMS is overriding the method here)?

agrare replied:
I thought if we renamed the method in `ExtManagementSystem` to `permitted_subclasses` that would free up the `supported_subclasses` name to be used here.  In `ExtManagementSystem` we can alias one to the other.

kbrock replied:
I don't see the need for this distinction but not my hill. done

Fryguy commented on spec/models/mixins/supports_feature_mixin_spec.rb:
```diff
@@ -200,4 +205,121 @@ def initialize(options = {})
+      expect(Post.subclasses_supporting(:delete).map(&:name)).to eq(%w[ProviderB::Post])
+      expect(Post.subclasses_supporting(:fake).map(&:name)).to match_array(%w[ProviderA::Post ProviderB::Post])
+    ensure
+      cleanup_subclass(Post)
```
Can this be done via a normal `stub_const` call?  If not should this be a general `after` block at the top?

kbrock replied:
so stub_const is successfully defining and cleaning up.
But rails caches the constants in the descendant tree.

After is tricky because we don't know what classes were defined?
I guess I could keep track of the classes defined and then clean those defined in the after test. But this is starting to get very picky and fragile.

kbrock replied:
https://github.com/rspec/rspec-mocks/issues/1240

So they are having issues with descendant parser as well. They were also suggesting using Class.new (which doesn't play as well with `ActiveRecord` since it doesn't have a name but works ok if we set a table name, which we have to do with stub constant anyway.

kbrock commented:
Push: put in a mechanism to automatically clean up the descendants

@Fryguy think this should be all set

agrare commented:
@kbrock the CI failures look to be introduced by this PR, likely the `ExtManagementSystem.supporting(:catalog)` one

agrare commented:
Actually, the current test makes no sense to me.  I wonder if this might have actually fixed a bug that someone wrote the test around in the first place?

Both the `it "only returns generic with no providers"` and `it "returns orchestration template and generic" do` include an `"amazon"` catalog_item_type but no amazon provider was created in the tests.

agrare commented on app/models/service_template.rb:
```diff
@@ -95,7 +95,7 @@ def self.all_catalog_item_types
       }
 
-      ExtManagementSystem.supported_types_for_catalog
+      ExtManagementSystem.supporting(:catalog)
```
Ah no I see why the tests are failing, this gets back to what I was talking about re supported classes vs supported instances.

The previous `ExtManagementSystem.supported_types_for_catalog` returns supported classes, `      ExtManagementSystem.supporting(:catalog)` returns instances from supported classes so this now returns `[]` where the previous method returned a number of provider classes

kbrock replied:
Good eye. looks like we want `subclasses_supporting` rather than instances of ems that support.
Subtle.

Fryguy commented on spec/models/mixins/supports_feature_mixin_spec.rb:
```diff
@@ -200,4 +213,116 @@ def initialize(options = {})
+    define_supporting_class("#{module_name}::#{parent.name}", parent, supports_values)
+  end
+
+  # these descenants are stored in a cache in active support
```
```suggestion
  # these descendants are stored in a cache in active support
```

kbrock commented:
push: pretty sure I fixed the sporadic test failure introduced by adding and removing classes

kbrock commented:
pushed: fixed cops

agrare commented on app/models/miq_template.rb:
```diff
@@ -31,8 +31,7 @@ def scan_via_ems?
   def self.eligible_for_provisioning
-    types = descendants.select { |d| d.supports?(:provisioning) }.map(&:name)
-    where(:type => types).where.not(:ems_id => nil)
+    supporting(:provisioning).where.not(:ems_id => nil)
```
Not for this PR but just noticed this can use the `.active` scope from `VmOrTemplate` (just `where.not(:ems_id => nil)`)

kbrock commented:
push: using `MiqTemplate#active`

kbrock commented:
> Actually, the current test makes no sense to me. I wonder if this might have actually fixed a bug that someone wrote the test around in the first place?
> 
> Both the `it "only returns generic with no providers"` and `it "returns orchestration template and generic" do` include an `"amazon"` catalog_item_type but no amazon provider was created in the tests.

That was a sporadic failure that I introduced.
Our specs were defining a subclass that got into the dependency tree, rails cached it, and then the subclasses went away. Our code then got confused.

I fixed this.

---

Any thing else?
